frontend/src/metabase/admin/settings/components/Email/SMTPConnectionForm.tsx

```diff
@@ -0,0 +1,168 @@
+
+import * as MetabaseAnalytics from "metabase/lib/analytics";
+import {
+  sendTestEmail,
+  updateEmailSettings,
+  clearEmailSettings,
+} from "../../settings";
+
+import SettingsBatchForm from "../SettingsBatchForm";
+
+const BREADCRUMBS = [[t`Email`, "/admin/settings/email"], [t`SMTP`]];
+
+const SEND_TEST_BUTTON_STATES = {
+  default: t`Send test email`,
+  working: t`Sending...`,
+  success: t`Sent!`,
+};
+type ButtonStateType = keyof typeof SEND_TEST_BUTTON_STATES;
+
+interface FormRefType {
+  handleFormErrors: (error: Error) => void;
+  setFormErrors: (formErrors: any) => void;
+  setState: ({ formData, dirty }: { formData: object; dirty: boolean }) => void;
+}
+
+interface SMTPConnectionFormProps {
+  elements: SettingElement[];
+  settingValues: Settings;
+}
+
+export const SMTPConnectionForm = ({
+  elements,
+  settingValues,
+}: SMTPConnectionFormProps) => {
+  const [sendingEmail, setSendingEmail] = useState<ButtonStateType>("default");
+
+  const formRef = useRef<FormRefType>();
+  const isHosted = useSelector(getIsHosted);
+  const isPaidPlan = useSelector(getIsPaidPlan);
+  const isEmailConfigured = useSelector(getIsEmailConfigured);
+
+  const dispatch = useDispatch();
+
+  const handleClearEmailSettings = useCallback(async () => {
+    await dispatch(clearEmailSettings());
+    // NOTE: reaching into form component is not ideal
+
+    formRef.current?.setState({ formData: {}, dirty: false });
+  }, [dispatch]);
+
+  const handleUpdateEmailSettings = useCallback(
+    async formData => {
+      try {
+        await dispatch(updateEmailSettings(formData));
+
+        if (!isEmailConfigured) {
+          dispatch(push("/admin/settings/email"));
+        }
+      } catch (error: any) {
+        formRef.current?.setFormErrors(
+          formRef.current?.handleFormErrors(error),
+        );
+      }
+    },
+    [dispatch, isEmailConfigured],
+  );
+
+  const handleSendTestEmail = useCallback(
+    async e => {
+      e.preventDefault();
+
+      setSendingEmail("working");
+      // NOTE: reaching into form component is not ideal
+      formRef.current?.setFormErrors(null);
+
+      try {
+        await dispatch(sendTestEmail());
+        setSendingEmail("success");
+        MetabaseAnalytics.trackStructEvent(
+          "Email Settings",
+          "Test Email",
+          "success",
+        );
+
+        // show a confirmation for 3 seconds, then return to normal
+        setTimeout(() => setSendingEmail("default"), 3000);
+      } catch (error: any) {
+        MetabaseAnalytics.trackStructEvent(
+          "Email Settings",
+          "Test Email",
+          "error",
+        );
+        setSendingEmail("default");
+        // NOTE: reaching into form component is not ideal
+        formRef.current?.setFormErrors(
+          formRef.current?.handleFormErrors(error),
+        );
+      }
+    },
+    [dispatch],
+  );
+
+  if (isHosted) {
+    dispatch(push("/admin/settings/email"));
```

Comment:
It would be nice to avoid dispatching actions right in the render function because it can lead to unexpected behaviors, infinite loops. Wdyt about moving it to an effect here?